pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -684,128 +530,67 @@ pub mod pallet {
 			Ok((address, incoming_msg))
 		}
 
-		/// Iterates over the outbound messages stored in the queue and attempts
-		/// to process them.
-		///
-		/// If a message fails to process it is moved to the
-		/// `FailedOutboundMessages` storage so that it can be executed again
-		/// via the `process_failed_outbound_message` extrinsic.
-		fn service_outbound_message_queue(max_weight: Weight) -> Weight {
-			let mut weight_used = Weight::zero();
-
-			let mut processed_entries = Vec::new();
-
-			for (nonce, (domain, sender, message)) in OutboundMessageQueue::<T>::iter() {
-				processed_entries.push(nonce);
-
-				let weight =
-					match Self::process_message(domain.clone(), sender.clone(), message.clone()) {
-						Ok(post_info) => {
-							Self::deposit_event(Event::OutboundMessageExecutionSuccess {
-								nonce,
-								sender,
-								domain,
-								message,
-							});
-
-							post_info
-								.actual_weight
-								.expect("Message processing success already ensured")
-								// Extra weight breakdown:
-								//
-								// 1 read for the outbound message
-								// 1 write for the event
-								// 1 write for the outbound message removal
-								.saturating_add(T::DbWeight::get().reads_writes(1, 2))
-						}
-						Err(e) => {
-							Self::deposit_event(Event::OutboundMessageExecutionFailure {
-								nonce,
-								sender: sender.clone(),
-								domain: domain.clone(),
-								message: message.clone(),
-								error: e.error,
-							});
-
-							FailedOutboundMessages::<T>::insert(
-								nonce,
-								(domain, sender, message, e.error),
-							);
-
-							e.post_info
-								.actual_weight
-								.expect("Message processing success already ensured")
-								// Extra weight breakdown:
-								//
-								// 1 read for the outbound message
-								// 1 write for the event
-								// 1 write for the failed outbound message
-								// 1 write for the outbound message removal
-								.saturating_add(T::DbWeight::get().reads_writes(1, 3))
-						}
-					};
-
-				weight_used = weight_used.saturating_add(weight);
-
-				if weight_used.all_gte(max_weight) {
-					break;
-				}
-			}
+		/// Sends the message to the `InboundMessageHandler`.
+		fn process_inbound_message(
+			domain_address: DomainAddress,
+			message: T::Message,
+		) -> (DispatchResult, Weight) {
+			let weight = Weight::from_parts(0, T::Message::max_encoded_len() as u64);
 
-			for entry in processed_entries {
-				OutboundMessageQueue::<T>::remove(entry);
+			match T::InboundMessageHandler::handle(domain_address, message) {
+				Ok(_) => (Ok(()), weight),
+				Err(e) => (Err(e), weight),
 			}
```

Comment:
The computation of `T::InboundMessageHandler::handle(domain_address, message)` is not counted in the `weight` and it should. Note that the inner computation of `T::InboundMessageHandler` can be a lot. i.e. if the message is something investment-related it can imply changes in foreign investment, opening swap orders, investments, etc...

Nevertheless, there is no easy solution here. I would add manually an extra weight. I would say 200_000_000 and 4096 of PoV can be enough for all cases by now. Interesting in your thoughts @wischli

Reply:
I would add that extra weight to the 2 extrinsics of the LP gateway queue that can end up processing this message when calling `MessageProcessor::process`. The only other place where this is called is `on_idle` but in that case we use the weight that's returned.

Reply:
Done.

Reply:
But we still skip the weight on the on_idle, right?

- `on_idle` uses the weight from `T::MessageProcessor::process`,  which only returns ` Weight::from_parts(0, T::Message::max_encoded_len() as u64)`. We're not taking into account the processed message weight. I think we should add the above weights to `process_inbound_message` as well. To get `on_idle` to compute correctly the weights

Reply:
Oh, right, for inbound messages we only return the `MaxEncodedLen` for the message, I will add the defensive weight in there as well.

Reply:
@wischli, this is the thread about the weights

Reply:
The weight is something we need to discuss; I thought `5_000_000_000` was excessively high for what, in my mind, is the worst case. Although probably `200_000_000` is pretty low. But I'm choosing the `5_000_000_000`

Reply:
@lemunozm IMO being defensive is desired and not an issue for now since we don't expect LP messages to explode in cardinality for the time being. As long as we don't have somewhat of more realistic overestimate via benchmarks, this value seems feasible to me, especially since it has not caused trouble so far. `5_000_000_000` computational weight equals 1% of the maximum we allow. Does not feel too defensive to me that we can easily fit processing 50+ messages into, at least, the majority of our blocks. WDYT @lemunozm?

Reply:
Though knowing that we could fit 1000-1500 transfers into a block if batched (bottleneck is signature verification if submitted in isolation),  I suppose we could lower it by half. Nevertheless, I cast my vote for the safer path.

Reply:
It makes sense what you say above! Let's be super defensive by now 👍🏻